datalad/interface/run.py

```diff
@@ -0,0 +1,268 @@
+                    r.pop('status', None)
+                    to_unlock.append(r)
+            if to_unlock:
+                for r in ds.unlock(to_unlock, return_type='generator', result_xfm=None):
+                    yield r
+        else:
+            # not a rerun, figure out where we are running
+            pwd = getpwd()
+            rel_pwd = relpath(pwd, start=ds.path)
+            if rel_pwd.startswith(pardir):
+                lgr.warning('Process working directory not inside the dataset, command run may not be reproducible.')
+                rel_pwd = None
+
+        # anticipate quoted compound shell commands
+        cmd = cmd[0] if isinstance(cmd, list) and len(cmd) == 1 else cmd
+
+        # TODO do our best to guess which files to unlock based on the command string
+        #      in many cases this will be impossible (but see --rerun). however,
+        #      generating new data (common case) will be just fine already
+
+        # we have a clean dataset, let's run things
+        cmd_exitcode = None
+        runner = Runner(cwd=pwd)
+        try:
+            lgr.info("== Command start (output follows) =====")
+            runner.run(
+                cmd,
+                # immediate output
+                log_online=True,
+                # not yet sure what we should do with the command output
+                # IMHO `run` itself should be very silent and let the command talk
+                log_stdout=False,
+                log_stderr=False,
+                expect_stderr=True,
+                expect_fail=True,
+                # TODO stdin
+            )
+        except CommandError as e:
+            # strip our own info from the exception. The original command output
+            # went to stdout/err -- we just have to exitcode in the same way
+            cmd_exitcode = e.code
+            if not rerun or rec_exitcode != cmd_exitcode:
+                # we failed during a fresh run, or in a different way during a rerun
+                # the latter can easily happen if we try to alter a locked file
+                #
+                # let's fail here, the command could have had a typo or some
+                # other undesirable condition. If we would `add` nevertheless,
+                # we would need to rerun and aggregate annex content that we
+                # likely don't want
+                # TODO add switch to ignore failure (some commands are stupid)
+                # TODO add the ability to `git reset --hard` the dataset tree on failure
+                # we know that we started clean, so we could easily go back, needs gh-1424
+                # to be able to do it recursively
+                raise CommandError(code=cmd_exitcode)
+
+        lgr.info("== Command exit (modification check follows) =====")
+
+        # ammend commit message with `run` info:
+        # - pwd if inside the dataset
+        # - the command itself
+        # - exit code of the command
+        run_info = {
```

Comment:
might be worth making it an ordered dict to guarantee the consistent order of the fields